File: 0110-balanced-binary-tree/0110-balanced-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int height(TreeNode* root)
    {
         if(root==NULL)
            return 0;   
         int lefty = height(root->left);
         if(lefty == -1)
            return -1;
         int righty = height(root->right);
         if(righty == -1)
            return -1;
         if(abs(lefty-righty)>1)
             return -1;   
         return 1+max(lefty,righty) ;      
    }
    bool isBalanced(TreeNode* root) {
        return (height(root) != -1);           
    }
};

File: 0110-balanced-binary-tree/0110-balanced-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        def helper(root):
            if not root:
                return True , 0
            lefty , hl = helper(root.left)
            if not lefty:
                return False , 0
            righty , rl = helper(root.right)    
            if not righty or abs(rl-hl)>1 :
                return False , 0
            return True , max(rl , hl)+1
        ans , _ = helper(root)
        return ans         

            



                


        